### ISSUE TITLE
pytest -m core peaks at 6.35 GB against ~7 GB CI runners

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#732]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

## Summary

The `pytest -m core` bucket — the **required** CI gate — peaks at **6.35 GB RSS**
against GitHub Linux runners with ~7 GB. Roughly 0.65 GB of headroom on the check that
gates every PR.

## Evidence

Measured with `/usr/bin/time -v` while landing #729:

- whole `-m core` suite: **6.35 GB** peak, 1151 passed, 8 skipped
- `tests/conftest.py`'s own threshold-gated `jax.clear_caches` fires at **6 GB**, and its
  tuning note (from #386) records **5.6 GB for a 735-test core bucket**

The bucket is now **1159 collected tests**, not 735. The threshold was tuned against a
bucket roughly 60% the current size, and the headroom has been quietly absorbed since.

## Why file it

This is **pre-existing and not caused by #729** — that PR *lowers* core's footprint by
moving its own 8.4 GB adjoint tests out of the bucket (8.4 GB → 3.62 GB for that file).
But the measurement surfaced how little slack is left, and the failure mode is
unpleasant: an OOM-killed runner on a required check looks like a flaky infrastructure
failure rather than a capacity problem, so it tends to get re-run rather than diagnosed.

Note this is a *different axis* from #566's compile wall and from the χ-recompile issue
in #460 — it is the aggregate resident set of the whole required bucket, not one
kernel's compile.

## Possible directions (unprioritised)

1. **Re-tune the `clear_caches` threshold** against the current bucket size, rather than
   leaving it at a value chosen for a 735-test bucket.
2. **Split `core`** into two CI jobs. Cheapest in engineering terms; costs a runner.
3. **Track peak RSS in CI** and fail on regression, so the next 0.5 GB is noticed when it
   lands rather than when a runner dies.
4. Identify the top-N memory offenders in `core` and decide individually whether they
   belong in the required gate at all — #729's adjoint tests are one example of a test
   that clearly did not.

(3) is probably the highest value regardless of which of the others is chosen: the
underlying problem is that nothing currently measures this, so it drifts.

**Comment by cderinbogaz** [comment]
Disclosure: I work on [Runnerhut](https://runnerhut.com). For the 6.35 GB peak, I would record RSS over the full required bucket and around each module, not just its maximum. That shows whether the 6 GB cleanup trigger acts early enough and whether releasing JAX caches actually lowers resident memory. Keep the full required test set as the control before deciding to move tests out. If additional memory is still needed, a matched Runnerhut Linux host-job comparison could be useful, subject to dependency/resource checks; more capacity should not replace tracking the growth.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Measured, because this issue's own top recommendation was that nothing measures it.

## The number

`main` @ `589f77e`, **CI's exact invocation** (`uv run pytest tests/ -m "core" --cov=tenax --cov-report=xml --cov-report=term-missing -v`), with `JAX_COMPILATION_CACHE_DIR` pointed at an empty directory so the persistent cache could not hide the peak:

```
Maximum resident set size:  7,239,884 KB  =  6.90 GB
Elapsed (wall clock):       1:00:47
Result:                     2070 passed, 9 skipped, 1935 deselected
```

Against this issue's baseline of **6.35 GB / 1159 tests** on 2026-07-31: **+0.55 GB (+8.7%)** while the bucket grew to **2070** tests. It drifted, exactly as predicted.

Note this is the *post-mitigation* peak — `conftest.py`'s threshold-gated `jax.clear_caches` fires during the run. And that threshold is still **6 GB**, tuned when the bucket was **735** tests. It is now nearly 3x stale and sits *below* the steady-state footprint, so it is firing routinely rather than as an exception.

## The premise is wrong, and it changes the severity

This issue says "**~7 GB** GitHub Linux runners" and reads the 6.35 GB as ~0.65 GB of headroom. That is the **2-core/7 GB** spec. `tenax-lab/tenax` is **public** and the required jobs run on `ubuntu-latest`, and GitHub moved public-repo Linux runners to **4-core/16 GB** in January 2024.

Corroborated empirically rather than only from the spec: the gate has been passing continuously while growing to a 6.90 GB peak. Against a real 7 GB ceiling — minus OS and runner-agent overhead — it would be OOM-killing constantly. It is not.

So utilisation is roughly **43%**, not 99%. **The imminent-OOM framing was never right, and is not right now.**

## What I think should happen

**Do not close this as fixed** — nothing fixed it, and #935 (which shed 24% of the gate's wall-clock) explicitly did not touch this axis: its fixture shrinks say *"chi is unchanged; only the sweep count is cut"*, and chi is what drives the resident set.

**Re-scope it instead**, keeping the part that was right:

1. **Correct the runner premise** (this comment). The urgency was based on a stale hardware spec.
2. **Re-tune the `clear_caches` threshold.** 6 GB for a 735-test bucket is now below the working set of a 2070-test one, so the mitigation is running hot. This is the only concrete defect the measurement found.
3. **Keep recommendation (3), track peak RSS in CI.** Still the highest-value item, and still unimplemented — it is why a 0.55 GB drift went unnoticed for five weeks. But it is now drift-monitoring, not OOM-prevention.

Recommendation (2) from the original list — splitting `core` into two jobs — is worth reconsidering, though on the **wall-clock** axis rather than this one: see #933, where the gate cleared the merge queue's 2-hour limit by 11 minutes on #935 and by ~4 seconds on #920.

## Caveats

Measured on a 251 GB / many-core box, not a 4-core runner. Peak RSS should be comparable since XLA-CPU allocates on demand, but anything scaling with core count could differ. The wall clock certainly does — 60 minutes here against 88-112 in CI — so treat the time as this machine's, not CI's.
